maro/simulator/scenarios/finance/sub_engines/stock/stock_business_engine.py

```diff
@@ -0,0 +1,270 @@
+                        )  # list  index is in action # self.snapshot
+                ret = TradeResult(
+                    self.name, action.item_index, actual_volume, tick, actual_price,
+                    commission_charge, is_success, is_trigger
+                )
+                if not is_trigger:
+                    if action.life_time != 1:
+                        action.life_time -= 1
+                        self._event_buffer.gen_atom_event(tick + 1, DecisionEvent, action)
+                        if action.id not in self._pending_orders:
+                            self._pending_orders.append(action.id)
+                    else:
+                        if action.id in self._pending_orders:
+                            self._pending_orders.remove(action.id)
+                            action.state = ActionState.expired
+                            action.finish_tick = tick
+                else:
+                    if action.id in self._pending_orders:
+                        self._pending_orders.remove(action.id)
+                    if is_success:
+                        action.state = ActionState.success
+                        if actual_volume > 0:
+                            self._stock_list[asset].average_cost = (
+                                (self._stock_list[asset].account_hold_num * self._stock_list[asset].average_cost) +
+                                (actual_price * actual_volume)
+                            ) / (self._stock_list[asset].account_hold_num + actual_volume)
+                        self._stock_list[asset].account_hold_num += actual_volume
+                    else:
+                        action.state = ActionState.failed
+                    action.finish_tick = tick
+                    action.action_result = ret
+            else:
+                print(
+                    "Warning: out of action scope and not allow split!", self.name,
+                    self._action_scope(action.action_type, action.item_index), action.number
+                )
+                action.state = ActionState.failed
+                action.finish_tick = tick
+
+        else:
+            # order canceled
+            self._canceled_orders.remove(action.id)
+            action.state = ActionState.canceled
+            action.finish_tick = tick
+        print("result", ret)
+        return ret
+
+    def cancel_order(self, action: Action):
+        if action.id in self._pending_orders:
+            self._pending_orders.remove(action.id)
+        if action.id not in self._canceled_orders:
+            print(f'Order canceled :{action.id}')
+            self._canceled_orders.append(action.id)
+
+    def reset(self):
+        for _, reader in self._readers.items():
+            reader.reset()
+
+        self._frame.reset()
+        self._snapshots.reset()
+        self._pending_orders = []
+        self._canceled_orders = []
+
+    def _action_scope(self, action_type: ActionType, stock_index: int):
```

Comment:
define and use class to these action scopes

Reply:
done